src/auth/oauth2/qgsauthoauth2edit.cpp

```diff
@@ -0,0 +1,925 @@
+    {
+      QgsDebugMsg( QStringLiteral( "FAILED to delete temp token cache file after copy: %1" ).arg( localcachefile ) );
+      return;
+    }
+  }
+}
+
+// slot
+void QgsAuthOAuth2Edit::tabIndexChanged( int indx )
+{
+  mCurTab = indx;
+  validateConfig();
+}
+
+// slot
+void QgsAuthOAuth2Edit::populateGrantFlows()
+{
+  cmbbxGrantFlow->addItem( QgsAuthOAuth2Config::grantFlowString( QgsAuthOAuth2Config::AuthCode ),
+                           static_cast<int>( QgsAuthOAuth2Config::AuthCode ) );
+  cmbbxGrantFlow->addItem( QgsAuthOAuth2Config::grantFlowString( QgsAuthOAuth2Config::Implicit ),
+                           static_cast<int>( QgsAuthOAuth2Config::Implicit ) );
+  cmbbxGrantFlow->addItem( QgsAuthOAuth2Config::grantFlowString( QgsAuthOAuth2Config::ResourceOwner ),
+                           static_cast<int>( QgsAuthOAuth2Config::ResourceOwner ) );
+}
+
+// slot
+void QgsAuthOAuth2Edit::definedCustomDirChanged( const QString &path )
+{
+  QFileInfo pinfo( path );
+  bool ok = pinfo.exists() || pinfo.isDir();
+
+  leDefinedDirPath->setStyleSheet( ok ? "" : QgsAuthGuiUtils::redTextStyleSheet() );
+
+  if ( ok )
+  {
+    loadDefinedConfigs();
+  }
+}
+
+// slot
+void QgsAuthOAuth2Edit::setCurrentDefinedConfig( const QString &id )
+{
+  mDefinedId = id;
+  QgsDebugMsg( QStringLiteral( "Set defined ID: %1" ).arg( id ) );
+  validateConfig();
+}
+
+void QgsAuthOAuth2Edit::currentDefinedItemChanged( QListWidgetItem *cur, QListWidgetItem *prev )
+{
+  Q_UNUSED( prev )
+
+  QgsDebugMsg( QStringLiteral( "Entered" ) );
+
+  QString id = cur->data( Qt::UserRole ).toString();
+  if ( !id.isEmpty() )
+  {
+    setCurrentDefinedConfig( id );
+  }
+}
+
+// slot
+void QgsAuthOAuth2Edit::selectCurrentDefinedConfig()
+{
+  if ( mDefinedId.isEmpty() )
+  {
+    return;
+  }
+
+  if ( !onDefinedTab() )
+  {
+    tabConfigs->setCurrentIndex( definedTab() );
+  }
+
+  lstwdgDefinedConfigs->selectionModel()->clearSelection();
+
+  for ( int i = 0; i < lstwdgDefinedConfigs->count(); ++i )
+  {
+    QListWidgetItem *itm = lstwdgDefinedConfigs->item( i );
+
+    if ( itm->data( Qt::UserRole ).toString() == mDefinedId )
+    {
+      lstwdgDefinedConfigs->setCurrentItem( itm, QItemSelectionModel::Select );
+      break;
+    }
+  }
+}
+
+// slot
+void QgsAuthOAuth2Edit::getDefinedCustomDir()
+{
+  QString extradir = QFileDialog::getExistingDirectory( this, tr( "Select extra directory to parse" ),
+                     QDir::homePath(), QFileDialog::ShowDirsOnly | QFileDialog::DontResolveSymlinks );
+  this->raise();
+  this->activateWindow();
+
+  if ( extradir.isEmpty() )
+  {
+    return;
+  }
+  leDefinedDirPath->setText( extradir );
+}
+
+void QgsAuthOAuth2Edit::initConfigObjs()
+{
+  mOAuthConfigCustom = new QgsAuthOAuth2Config( this );
+  mOAuthConfigCustom->setConfigType( QgsAuthOAuth2Config::Custom );
+  mOAuthConfigCustom->setToDefaults();
+}
+
+void QgsAuthOAuth2Edit::deleteConfigObjs()
+{
+  delete mOAuthConfigCustom;
```

Comment:
Candidate for unique_ptr

Reply:
@nyalldawson thanks for the review! I'm looking into this candidate for unique_ptr, it's a QObject so I think we could just leave memory management to Qt (it's parented) and completely drop the dtor, what do you think?

Reply:
@nyalldawson nevermind, I've done that in latest commit